File: ruby/lib/k256/leader_types.rb
# frozen_string_literal: true

module K256
  module Leader
    # Subscription channel constants
    CHANNEL_LEADER_SCHEDULE = "leader_schedule"
    CHANNEL_GOSSIP = "gossip"
    CHANNEL_SLOTS = "slots"
    CHANNEL_ALERTS = "alerts"
    ALL_CHANNELS = [CHANNEL_LEADER_SCHEDULE, CHANNEL_GOSSIP, CHANNEL_SLOTS, CHANNEL_ALERTS].freeze

    # Message kind — how to consume the message
    # "snapshot" = full state replacement
    # "diff" = merge into snapshot using key
    # "event" = append-only
    KINDS = %w[snapshot diff event].freeze

    # GossipPeer represents a single gossip network peer.
    GossipPeer = Struct.new(
      :identity, :tpu_quic, :tpu_udp, :tpu_forwards_quic, :tpu_forwards_udp,
      :tpu_vote, :gossip_addr, :version, :shred_version, :stake,
      :commission, :is_delinquent, :wallclock,
      :country_code, :continent_code, :asn, :as_name, :as_domain,
      keyword_init: true
    )

    SlotUpdate = Struct.new(:slot, :leader, :block_height, keyword_init: true)
    RoutingHealth = Struct.new(:leaders_total, :leaders_in_gossip, :leaders_missing_gossip,
                              :leaders_without_tpu_quic, :leaders_delinquent, :coverage,
                              keyword_init: true)
    GossipDiff = Struct.new(:timestamp_ms, :added, :removed, :updated, keyword_init: true)
    SkipEvent = Struct.new(:slot, :leader, :assigned, :produced, keyword_init: true)
    IpChange = Struct.new(:identity, :old_ip, :new_ip, :timestamp_ms, keyword_init: true)
    LeaderHeartbeat = Struct.new(:timestamp_ms, :current_slot, :connected_clients, :gossip_peers,
                                keyword_init: true)
  end
end

File: swift/Sources/K256/Leader/LeaderTypes.swift
import Foundation

/// Leader Schedule WebSocket message types.
/// All messages are JSON text frames with: type, kind, key (optional), data.

public enum LeaderChannel: String, CaseIterable {
    case leaderSchedule = "leader_schedule"
    case gossip = "gossip"
    case slots = "slots"
    case alerts = "alerts"
}

public enum MessageKind: String, Codable {
    case snapshot
    case diff
    case event
}

public struct GossipPeer: Codable {
    public let identity: String
    public let tpuQuic: String?
    public let tpuUdp: String?
    public let tpuForwardsQuic: String?
    public let tpuForwardsUdp: String?
    public let tpuVote: String?
    public let gossipAddr: String?
    public let version: String
    public let shredVersion: UInt16
    public let stake: UInt64
    public let commission: UInt8
    public let isDelinquent: Bool
    public let wallclock: UInt64
    /// ISO 3166 country code (e.g. "US", "DE")
    public let countryCode: String?
    /// Two-letter continent code (e.g. "NA", "EU")
    public let continentCode: String?
    /// ASN string (e.g. "AS15169")
    public let asn: String?
    /// AS organization name (e.g. "Google LLC")
    public let asName: String?
    /// AS organization domain (e.g. "google.com")
    public let asDomain: String?
}

public struct SlotUpdateData: Codable {
    public let slot: UInt64
    public let leader: String
    public let blockHeight: UInt64
}

public struct RoutingHealthData: Codable {
    public let leadersTotal: UInt32
    public let leadersInGossip: UInt32
    public let leadersMissingGossip: [String]
    public let leadersWithoutTpuQuic: [String]
    public let leadersDelinquent: [String]
    public let coverage: String
}

public struct GossipDiffData: Codable {
    public let timestampMs: UInt64
    public let added: [GossipPeer]
    public let removed: [String]
    public let updated: [GossipPeer]
}

public struct SkipEventData: Codable {
    public let slot: UInt64
    public let leader: String
    public let assigned: UInt32
    public let produced: UInt32
}

public struct IpChangeData: Codable {
    public let identity: String
    public let oldIp: String
    public let newIp: String
    public let timestampMs: UInt64
}

public struct LeaderHeartbeatData: Codable {
    public let timestampMs: UInt64
    public let currentSlot: UInt64
    public let connectedClients: UInt32
    public let gossipPeers: UInt32
}

public struct MessageSchemaEntry: Codable {
    public let type: String
    public let tag: String
    public let kind: String
    public let key: String?
    public let description: String
}

public struct LeaderSubscribedData: Codable {
    public let channels: [String]
    public let currentSlot: UInt64
    public let epoch: UInt64
    public let schema: [MessageSchemaEntry]
}
